### PR TITLE
Kafka Avro Sink with Editor

### PR DESCRIPTION
**Pull request by nadberezny**

### CODE REVIEW

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala [hunk 1]
```diff
@@ -3,23 +3,95 @@ package pl.touk.nussknacker.engine.avro.sink
...
+      val emptyObj = Map.empty[String, AnyRef]
+      lazyObjectFieldsSequence
+        .map { list =>
+          list.foldLeft(emptyObj) { case (obj, field) =>
+            obj + (field._1 -> field._2)
+          }
+        }
+    }
+
+    override def flatMap(value: Context, out: Collector[ValueWithContext[KeyedValue[AnyRef, AnyRef]]]): Unit = {
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala:51]: Let's use proper error handling... Like in EncodeAvroRecordFunction

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+          paramEither match {
+            case Right(value) => NextParameters(value :: Nil)
+            case Left(fields) => NextParameters(fields)
+          }
+        }
+      }.valueOr(e => FinalResults(context, e :: Nil))
+  }
+
+  /* TODO, FIXME:
+    1. optional params with default in avro schema
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala:108]: Let's add this comment to toParamEither method?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import pl.touk.nussknacker.engine.definition.parameter.editor.ParameterTypeEditorDeterminer
+
+object KafkaAvroSinkFactoryV2 {
+
+  private val restrictedParamNames = Set(SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, TopicParamName)
+
+  private def containsRestrictedNames(obj: TypedObjectTypingResult): Boolean =
+    (obj.fields.keySet & restrictedParamNames).nonEmpty
+
+  private def toParamEither(typing: TypingResult)(implicit nodeId: NodeId): Validated[ProcessCompilationError, Either[List[Parameter], Parameter]] = {
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala:27]: Why do we need this Either? Why Parameter is not just list with one element?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    def toParamEither(typing: TypingResult, isRoot: Boolean)(implicit nodeId: NodeId) =
+      typing match {
+        case typed.typing.Unknown =>
+          Invalid(CustomNodeError(nodeId.id, "Cannot determine typing for provided schema", None))
+        case typedObject: TypedObjectTypingResult if containsRestrictedNames(typedObject) =>
+          Invalid(CustomNodeError(nodeId.id, s"""Record field name is restricted. Restricted names are ${restrictedParamNames.mkString(", ")}""", None))
+        case typedObject: TypedObjectTypingResult if isRoot =>
+          Valid(Left(toObjectFieldParams(typedObject)))
+        case _: TypedObjectTypingResult =>
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala:48]: Can we add handling "simple" nested elements? Or at least add comment for that?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+import pl.touk.nussknacker.engine.api.{LazyParameter, MetaData}
+import pl.touk.nussknacker.engine.avro.schemaregistry.SchemaRegistryProvider
+import pl.touk.nussknacker.engine.avro.{KafkaAvroBaseTransformer, SchemaDeterminerErrorHandler}
+import pl.touk.nussknacker.engine.avro.KafkaAvroBaseTransformer.{SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, SinkValueParamName, TopicParamName}
+import pl.touk.nussknacker.engine.avro.typed.AvroSchemaTypeDefinitionExtractor
+import pl.touk.nussknacker.engine.flink.api.process.FlinkSink
+import pl.touk.nussknacker.engine.api.typed
+import pl.touk.nussknacker.engine.definition.parameter.editor.ParameterTypeEditorDeterminer
+
+object KafkaAvroSinkFactoryV2 {
```
[the remark this answers is not in the document]
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala]: done

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala [hunk 6]
```diff
@@ -3,23 +3,95 @@ package pl.touk.nussknacker.engine.avro.sink
...
+  import pl.touk.nussknacker.engine.api.{Context, LazyParameter, LazyParameterInterpreter, ValueWithContext}
+  import pl.touk.nussknacker.engine.flink.api.process.{FlinkLazyParameterFunctionHelper, LazyParameterInterpreterFunction}
+  import pl.touk.nussknacker.engine.flink.util.keyed
+  import pl.touk.nussknacker.engine.flink.util.keyed.KeyedValue
+
+  import scala.util.control.NonFatal
+
+
+  // TODO FIXME: sanitize key!
+  class KeyedObjectMapper(protected val lazyParameterHelper: FlinkLazyParameterFunctionHelper,
```
[the remark this answers is not in the document]
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala]: :+1:

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala [hunk 7]
```diff
@@ -28,9 +100,14 @@ class KafkaAvroSink(preparedTopic: PreparedKafkaTopic, versionOption: SchemaVers
   @transient final protected lazy val avroEncoder = BestEffortAvroEncoder(validationMode)
 
   override def registerSink(dataStream: DataStream[InterpretationResult], flinkNodeContext: FlinkCustomNodeContext): DataStreamSink[_] = {
-    dataStream
-      .map(_.finalContext)
-      .map(new KeyedValueMapper(flinkNodeContext.lazyParameterHelper, key, value))
+    val ds1 = dataStream.map(_.finalContext)
+    val ds2 = valueEither match {
+      case Right(value) =>
```
[the remark this answers is not in the document]
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala]: :+1:

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala (hunk 1 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala]: Done

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala (hunk 3 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala]: We need a distinction between object with one field and single value. I've replaced Either with new abstraction

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala (hunk 4 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala]: I've added handling of nested fields

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  private val restrictedParamNames = Set(SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, TopicParamName)
+
+  def apply(typing: TypingResult)(implicit nodeId: NodeId): Validated[ProcessCompilationError, AvroSinkValueParameter] =
+    toSinkValueParameter(typing, paramName = None, isTopLevel = true)
+
+  private def toSinkValueParameter(typing: TypingResult, paramName: Option[String], isTopLevel: Boolean)
+                                  (implicit nodeId: NodeId): Validated[ProcessCompilationError, AvroSinkValueParameter] =
+    typing match {
+      case typed.typing.Unknown =>
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala:26]: Or just leave plain expression?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        Invalid(CustomNodeError(nodeId.id, "Cannot determine typing for provided schema", None))
+
+      case typedObject: TypedObjectTypingResult if containsRestrictedNames(typedObject) =>
+        Invalid(CustomNodeError(nodeId.id, s"""Record field name is restricted. Restricted names are ${restrictedParamNames.mkString(", ")}""", None))
+
+      /* kafka-avro-serializer does not support Array at top level
+         [https://github.com/confluentinc/schema-registry/issues/1298] */
+      case TypedClass(clazz, _) if isTopLevel && clazz == classOf[java.util.List[_]] =>
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala:34]: Should it be handled here? We already have TypingResult from schema from schema regitry, so maybe we should just assume it's sth schema registry can handle?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  def apply(typing: TypingResult)(implicit nodeId: NodeId): Validated[ProcessCompilationError, AvroSinkValueParameter] =
+    toSinkValueParameter(typing, paramName = None, isTopLevel = true)
+
+  private def toSinkValueParameter(typing: TypingResult, paramName: Option[String], isTopLevel: Boolean)
+                                  (implicit nodeId: NodeId): Validated[ProcessCompilationError, AvroSinkValueParameter] =
+    typing match {
+      case typed.typing.Unknown =>
+        Invalid(CustomNodeError(nodeId.id, "Cannot determine typing for provided schema", None))
+
+      case typedObject: TypedObjectTypingResult if containsRestrictedNames(typedObject) =>
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala:29]: Can we add TODO to handle this case somehow?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+  private def containsRestrictedNames(obj: TypedObjectTypingResult): Boolean = {
+    val fieldNames = obj.fields.keySet
+    fieldNames.nonEmpty & (fieldNames & restrictedParamNames).nonEmpty
+  }
+
+  private def sequence(l: List[(FieldName, Validated[ProcessCompilationError, AvroSinkValueParameter])])
+  : Validated[ProcessCompilationError, List[(FieldName, AvroSinkValueParameter)]] = {
+    val zero: Validated[ProcessCompilationError, List[(FieldName, AvroSinkValueParameter)]] = Valid(Nil)
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala:59]: I think this method shows we should use ValidatedNel and applicative here, now we're returning only first error?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/BaseKafkaAvroSinkFactory.scala [hunk 12]
```diff
@@ -33,7 +33,7 @@ abstract class BaseKafkaAvroSinkFactory extends SinkFactory {
     val schemaUsedInRuntime = schemaDeterminer.toRuntimeSchema(schemaData)
 
     val clientId = s"${processMetaData.id}-${preparedTopic.prepared}"
-    new KafkaAvroSink(preparedTopic, version, key, value, kafkaConfig, serializationSchemaFactory,
+    new KafkaAvroSink(preparedTopic, version, key, AvroSinkPrimitiveValue(value), kafkaConfig, serializationSchemaFactory,
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/BaseKafkaAvroSinkFactory.scala:36]: Is this really primitive value? Maybe we should change name to sth like SingleValueParameter?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      }
+      lazyRecord.product(lazyParam).map { case (rec, value) =>
+        rec + (fieldName -> value)
+      }
+    }
+
+  override def flatMap(value: Context, out: Collector[ValueWithContext[KeyedValue[AnyRef, AnyRef]]]): Unit =
+    exceptionHandler.handling(Some(nodeId), value) {
+      out.collect(ValueWithContext(interpret(value), value))
+    }.orNull
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala:67]: why do we need this .orNull??

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala [hunk 14]
```diff
@@ -16,31 +15,42 @@ import pl.touk.nussknacker.engine.flink.api.process.{FlinkCustomNodeContext, Fli
...
 
   /**
    * Right now we support it incorrectly, because we don't use default sink behavior with expression..
    */
   override def testDataOutput: Option[Any => String] = Some(value => Option(value).map(_.toString).getOrElse(""))
 
+  private def toValueWithContext(ds: DataStream[Context], flinkNodeContext: FlinkCustomNodeContext): DataStream[ValueWithContext[KeyedValue[AnyRef, AnyRef]]] =
+    sinkValue match {
+      case AvroSinkPrimitiveValue(value) =>
+        ds.map(new KeyedValueMapper(flinkNodeContext.lazyParameterHelper, key, value))
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala:49]: Let's add TODO to KeyedValueMapper to change it also to flatMap and handle errors properly?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    Option(lazyParameterInterpreter).foreach(_.close())
+  }
+
+  private lazy val emptyRecord: LazyParameter[Map[String, AnyRef]] = lazyParameterInterpreter
+    .pure[Map[String, AnyRef]](Map.empty, typing.Typed[Map[String, AnyRef]])
+
+  private lazy val record: LazyParameter[Map[String, AnyRef]] =
+    merge(emptyRecord, sinkRecord)
+
+  private def merge(agg: LazyParameter[Map[String, AnyRef]], sinkRecord: AvroSinkRecordValue): LazyParameter[Map[String, AnyRef]] =
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala:53]: I wonder how it works performance-wise... let's add some TODO to check it...

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameterTest.scala [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.scalatest.{FunSuite, Matchers}
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.NodeId
+import pl.touk.nussknacker.engine.api.definition.{DualParameterEditor, Parameter, StringParameterEditor}
+import pl.touk.nussknacker.engine.api.editor.DualEditorMode
+import pl.touk.nussknacker.engine.api.typed.typing
+import pl.touk.nussknacker.engine.avro.KafkaAvroBaseTransformer.SinkValueParamName
+import pl.touk.nussknacker.engine.util.typing.TypingUtils
+
+class AvroSinkValueParameterTest extends FunSuite with Matchers {
```
Comment (mproch) [inline engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameterTest.scala:11]: Can you add some tests also for error cases?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValue.scala [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package pl.touk.nussknacker.engine.avro.sink
+
+import pl.touk.nussknacker.engine.api.LazyParameter
+
+object AvroSinkValue {
+
+  def applyUnsafe(sinkParameter: AvroSinkValueParameter, parameterValues: Map[String, Any]): AvroSinkValue =
+    sinkParameter match {
+      case AvroSinkPrimitiveValueParameter(param) =>
+        val value = parameterValues(param.name).asInstanceOf[LazyParameter[AnyRef]]
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValue.scala:10]: let's add some better error?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryWithEditor.scala [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+      valueParams, _) if valueParams.nonEmpty => FinalResults(context, Nil)
+  }
+
+  override def initialParameters: List[Parameter] = {
+    implicit val nodeId: NodeId = NodeId("")
+    val topic = getTopicParam.value
+    topic :: getVersionParam(Nil) :: paramsDeterminedAfterSchema
+  }
+
+  override def contextTransformation(context: ValidationContext, dependencies: List[NodeDependencyValue])
```
Comment (mproch) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryWithEditor.scala:83]: :+1:  nice idea, now we can extract/share some logic with "raw" sink more easily

**Review by mproch — COMMENTED** [review]
Overall looks quite ok, I added some minor comments

File: engine/flink/generic/src/main/scala/pl/touk/nussknacker/genericmodel/GenericConfigCreator.scala [hunk 19]
```diff
@@ -45,10 +45,12 @@ class GenericConfigCreator extends EmptyProcessConfigCreator {
   override def sinkFactories(processObjectDependencies: ProcessObjectDependencies): Map[String, WithCategories[SinkFactory]] = {
     val schemaRegistryProvider = createSchemaProvider(processObjectDependencies)
     val kafkaAvroSinkFactory = new KafkaAvroSinkFactory(schemaRegistryProvider, processObjectDependencies)
+    val kafkaAvroSinkFactoryWithEditor = new KafkaAvroSinkFactoryWithEditor(schemaRegistryProvider, processObjectDependencies)
 
     Map(
       "kafka-json" -> defaultCategory(new GenericKafkaJsonSink(processObjectDependencies)),
       "kafka-avro" -> defaultCategory(kafkaAvroSinkFactory),
+      "kafka-avro-editor" -> defaultCategory(kafkaAvroSinkFactoryWithEditor),
```
[the remark this answers is not in the document]
Comment (nadberezny) [reply engine/flink/generic/src/main/scala/pl/touk/nussknacker/genericmodel/GenericConfigCreator.scala]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/efdd5e974b882878d44db537e43905ea3612621f

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala (hunk 8 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/3576cbaed9a20967214fb7c476409fac06ee4171

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala (hunk 10 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala:28]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/c9c38c070bd37ae7273f0f486f4a95700033e7bc

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala (hunk 14 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala:49]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/9f8cf216e127ebb7af131f51b2e1a2bec6283526

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala (hunk 15 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala:54]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/d779d8ebf469102dd0b3be4e56f5252955cbd6de

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameterTest.scala (hunk 16 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameterTest.scala:12]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/c502a5f0ed1fae5369bcdacfa85253d45f6f74b6

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValue.scala (hunk 17 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValue.scala]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/baba520d09e4a11b6c94cc49b95146a3d227d605

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala (hunk 11 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala]: done here: 
https://github.com/TouK/nussknacker/pull/1405/commits/f15c331bf3cdedc082358da86a6a7b9027f12e02

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/BaseKafkaAvroSinkFactory.scala (hunk 12 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/BaseKafkaAvroSinkFactory.scala]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/9c6c765331d9c50413c687cc6f30b044aeeac7d8

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala (hunk 13 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KeyedRecordFlatMapper.scala]: :+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/6ad3f15177f18472c5dbebfe47e70c7967c99888

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala (hunk 9 above)
Comment (nadberezny) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala]: :+1:  done
https://github.com/TouK/nussknacker/pull/1405/commits/88e3672307aa1eedbbf4819b6fc8dd86cd63c1f6

**Review by mproch — APPROVED** [review]
:+1: I think it's ok to merge, thanks :)

**Pull request merged**